rust/lancedb/src/catalog/listing.rs

```diff
@@ -0,0 +1,560 @@
+                }
+
+                Ok(Self {
+                    uri: String::from(url.clone()),
+                    query_string: url.query().map(|s| s.to_string()),
+                    base_path,
+                    object_store,
+                    store_wrapper: None,
+                    storage_options,
+                })
+            }
+            Err(_) => Self::open_path(uri).await,
+        }
+    }
+
+    fn database_path(&self, name: &str) -> ObjectStorePath {
+        self.base_path.child(name)
+    }
+}
+
+#[async_trait]
+impl Catalog for ListingCatalog {
+    async fn database_names(&self, request: DatabaseNamesRequest) -> Result<Vec<String>> {
+        let mut f = self
+            .object_store
+            .read_dir(self.base_path.clone())
+            .await?
+            .iter()
+            .map(Path::new)
+            .filter_map(|p| p.file_stem().and_then(|s| s.to_str().map(String::from)))
+            .collect::<Vec<String>>();
+        f.sort();
+
+        if let Some(start_after) = request.start_after {
+            let index = f
+                .iter()
+                .position(|name| name.as_str() > start_after.as_str())
+                .unwrap_or(f.len());
+            f.drain(0..index);
+        }
+        if let Some(limit) = request.limit {
+            f.truncate(limit as usize);
+        }
+        Ok(f)
+    }
+
+    async fn create_database(&self, request: CreateDatabaseRequest) -> Result<Arc<dyn Database>> {
+        let db_path = self.database_path(&request.name);
+        let db_path_str = to_local_path(&db_path);
+        let exists = Path::new(&db_path_str).exists();
+
+        match request.mode {
+            CreateDatabaseMode::Create if exists => {
+                return Err(Error::DatabaseAlreadyExists { name: request.name })
+            }
+            CreateDatabaseMode::Create => {
+                let db_local_path = to_local_path(&db_path);
+                create_dir_all(&db_local_path).unwrap();
+            }
+            CreateDatabaseMode::ExistOk => {
+                if !exists {
+                    let db_local_path = to_local_path(&db_path);
+                    create_dir_all(&db_local_path).unwrap();
+                }
+            }
+            CreateDatabaseMode::Overwrite => {
+                if exists {
+                    self.drop_database(&request.name).await?;
+                }
+                let db_local_path = to_local_path(&db_path);
+                create_dir_all(&db_local_path).unwrap();
+            }
+        }
+
+        let db_uri = format!("/{}/{}", self.base_path, request.name);
+
+        let connect_request = ConnectRequest {
+            uri: db_uri,
+            api_key: None,
+            region: None,
+            host_override: None,
+            #[cfg(feature = "remote")]
+            client_config: Default::default(),
+            read_consistency_interval: None,
+            storage_options: HashMap::new(),
```

Comment:
How should we populate `storage_options` for a database?  Should we copy the `storage_options` from the catalog?  Or should we pass in the `database_options` from the request?

Reply:
> Should we copy the storage_options from the catalog?
I'd prefer this option.